odoo/addons/base/models/ir_attachment.py

```diff
@@ -50,21 +90,26 @@ def _storage(self):
     def _filestore(self):
         return config.filestore(self._cr.dbname)
 
+    @api.model
+    def storage_locations(self):
+        return ['db', 'file']
+
     @api.model
     def force_storage(self):
-        """Force all attachments to be stored in the currently configured storage"""
         if not self.env.is_admin():
             raise AccessError(_('Only administrators can execute this action.'))
-
-        # domain to retrieve the attachments to migrate
-        domain = {
-            'db': [('store_fname', '!=', False)],
-            'file': [('db_datas', '!=', False)],
-        }[self._storage()]
-
-        for attach in self.search(domain):
-            attach.write({'datas': attach.datas})
-        return True
+        # Migrate only binary attachments and make sure to miragte those that belong 
+        # to a field as well as attachments that are not assigned to a field. 
+        self.search(expression.AND([self._get_storage_domain(self._storage()), [
+            '&', ('type', '=', 'binary'), '|', ('res_field', '=', False), ('res_field', '!=', False) 
+        ]])).migrate()
+
+    def migrate(self):
+        record_count = len(self)
+        storage = self._storage().upper()
+        for index, attach in enumerate(self):
+            _logger.info(_("Migrate Attachment %s of %s to %s") % (index + 1, record_count, storage))
+            attach.write({'datas': attach.datas, 'mimetype': attach.mimetype})
```

Comment:
This is necessary because otherwise a lot of mimetypes will break. The value `datas` triggers that the mimetypes are evaluated again, although nothing has changed. But now as only `datas` is given as a value so the recalculation is wrong in many cases.